### PR TITLE
Add patch to remove `block.*` settings from LVM and Ceph RBD block volumes

### PR DESCRIPTION
**Pull request by roosterfish**
This adds a patch to remove the `block.*` settings from VMs and custom volumes of type block in the LVM and Ceph driver. 

That is required after https://github.com/canonical/lxd/pull/12805 and follows the same procedure as in https://github.com/canonical/lxd/pull/12390.

### CODE REVIEW

**Comment by roosterfish** [comment]
I have removed https://github.com/canonical/lxd/pull/12390/files#diff-967760ebb0c0b53b1eb0f2c1f141563885405f50c26d77dab588b90d26528b0aR1087 in order to be able to patch Ceph RBD backed volumes. 
I guess the patch is running locally on each node so this would try to apply the patch on matching Ceph volumes x times (if clustered). Will this cause any race?

**Comment by tomponline** [comment]
> I have removed https://github.com/canonical/lxd/pull/12390/files#diff-967760ebb0c0b53b1eb0f2c1f141563885405f50c26d77dab588b90d26528b0aR1087 in order to be able to patch Ceph RBD backed volumes. I guess the patch is running locally on each node so this would try to apply the patch on matching Ceph volumes x times (if clustered). Will this cause any race?

Yes it will. For Ceph I suggest we need to detect if the member is leader and only run it on the leader.

**Comment by roosterfish** [comment]
@tomponline looks to be ok now, the code for checking the leader was just a few lines above :)

File: lxd/patches.go [hunk 1]
```diff
@@ -1174,4 +1175,141 @@ func patchStorageZfsUnsetInvalidBlockSettingsV2(_ string, d *Daemon) error {
...
+			if err != nil {
+				return fmt.Errorf("Failed getting storage pool ID of pool %q: %w", pool, err)
+			}
+
+			driverName, err := tx.GetStoragePoolDriver(ctx, poolID)
+			if err != nil {
+				return fmt.Errorf("Failed getting storage pool driver of pool %q: %w", pool, err)
+			}
+
+			if !shared.ValueInSlice[string](driverName, []string{"lvm", "ceph"}) {
```
Comment (tomponline) [inline lxd/patches.go:1213]: @roosterfish do we need a fix for dir or btrfs?

Comment (roosterfish) [reply lxd/patches.go:1213]: They already block creation of such volumes:

```bash
ulian@thinkpad:~$ lxc storage create btrfs btrfs
Storage pool btrfs created
julian@thinkpad:~$ lxc storage volume create btrfs vol --type block block.filesystem=xfs
Error: Invalid option for volume "default_vol" option "block.filesystem"
julian@thinkpad:~$ lxc storage create dir dir
Storage pool dir created
julian@thinkpad:~$ lxc storage volume create dir vol --type block block.filesystem=xfs
Error: Invalid option for volume "default_vol" option "block.filesystem"

```

File: lxd/patches.go [hunk 2]
```diff
@@ -1174,4 +1175,141 @@ func patchStorageZfsUnsetInvalidBlockSettingsV2(_ string, d *Daemon) error {
...
+		return err
+	})
+	if err != nil {
+		return err
+	}
+
+	leaderAddress, err := d.gateway.LeaderAddress()
+	if err != nil {
+		// If we're not clustered, we're the leader.
+		if errors.Is(err, cluster.ErrNodeIsNotClustered) {
```
Comment (tomponline) [inline lxd/patches.go:1252]: this looks wrong, we shouldnt return an error if we are not clustered?

Comment (tomponline) [reply lxd/patches.go:1252]: You can also use `s.ServerClustered` for this check too.

Comment (roosterfish) [reply lxd/patches.go:1252]: Ah yes that comes from trying to fix what the linter complained about. I didn't negate the condition. Let me fix this.

File: lxd/patches.go [hunk 3]
```diff
@@ -1174,4 +1175,141 @@ func patchStorageZfsUnsetInvalidBlockSettingsV2(_ string, d *Daemon) error {
...
+	} else if localConfig.ClusterAddress() == leaderAddress {
+		isLeader = true
+	}
+
+	var volType int
+
+	for pool, volumes := range poolVolumes {
+		for _, vol := range volumes {
+			// Load storage pool.
+			p, err := storagePools.LoadByName(s, poolIDNameMap[pool])
```
Comment (tomponline) [inline lxd/patches.go:1266]: Loading the pool for every volume in the pool seems inefficient to me.

File: lxd/patches.go [hunk 4]
```diff
@@ -1174,4 +1175,141 @@ func patchStorageZfsUnsetInvalidBlockSettingsV2(_ string, d *Daemon) error {
...
+			if err != nil {
+				return fmt.Errorf("Failed loading pool %q: %w", poolIDNameMap[pool], err)
+			}
+
+			// Ensure the renaming is done only on the cluster leader for remote storage pools.
+			if p.Driver().Info().Remote && !isLeader {
```
Comment (tomponline) [inline lxd/patches.go:1272]: Checking this for every volume even though its pool specific seems inefficient to me

File: lxd/patches.go (hunk 3 above)
Comment (roosterfish) [reply lxd/patches.go]: Thanks, fixed.

File: lxd/patches.go (hunk 4 above)
Comment (roosterfish) [reply lxd/patches.go]: Thanks, fixed.

File: lxd/patches.go [hunk 5]
```diff
@@ -1174,4 +1175,146 @@ func patchStorageZfsUnsetInvalidBlockSettingsV2(_ string, d *Daemon) error {
...
+		return err
+	}
+
+	if !d.serverClustered {
+		// If we're not clustered, we're the leader.
+		isLeader = true
+	} else {
+		leaderAddress, err := d.gateway.LeaderAddress()
+		if err != nil {
+			return nil
```
Comment (tomponline) [inline lxd/patches.go:1212]: this looks problematic, if there is an issue getting the leader address, we permanently skip applying the patch to this member's volumes?

Comment (roosterfish) [reply lxd/patches.go]: Good catch thanks! It should definitely return the actual error message. Must be a leftover from trying to escape the loop in earlier versions.

Comment (roosterfish) [reply lxd/patches.go]: That is fixed now.

File: lxd/patches.go [hunk 6]
```diff
@@ -1174,4 +1175,146 @@ func patchStorageZfsUnsetInvalidBlockSettingsV2(_ string, d *Daemon) error {
...
+
+		return fmt.Errorf("Failed getting storage pool names: %w", err)
+	}
+
+	// Get the current leader.
+	var localConfig *node.Config
+	isLeader := false
+
+	err = d.db.Node.Transaction(s.ShutdownCtx, func(ctx context.Context, tx *db.NodeTx) error {
+		localConfig, err = node.ConfigLoad(ctx, tx)
```
Comment (tomponline) [inline lxd/patches.go:1199]: Can just use s.LocalConfig.ClusterAddress() for this, but please test.

Comment (roosterfish) [reply lxd/patches.go]: Thanks for this hint, wasn't aware of it. Looks to be working fine for both clustered and non-clustered setups.

**Review by tomponline — APPROVED** [review]

**Pull request merged**